pkg/doc/jwt/jwt.go

```diff
@@ -0,0 +1,271 @@
+	}
+
+	return parseJWS(jwt, pOpts)
+}
+
+// DecodeClaims fills input c with claims of a token.
+func (j *JSONWebToken) DecodeClaims(c interface{}) error {
+	pBytes, err := json.Marshal(j.Payload)
+	if err != nil {
+		return err
+	}
+
+	return json.Unmarshal(pBytes, c)
+}
+
+// LookupStringHeader makes look up of particular header with string value.
+func (j *JSONWebToken) LookupStringHeader(name string) string {
+	if headerValue, ok := j.Headers[name]; ok {
+		if headerStrValue, ok := headerValue.(string); ok {
+			return headerStrValue
+		}
+	}
+
+	return ""
+}
+
+// SerializeSigned makes (compact) serialization of token.
+func (j *JSONWebToken) SerializeSigned(signer jose.Signer, detached bool) (string, error) {
+	return j.serialize(signer, detached)
+}
+
+// SerializeUnsecured build unsecured JWT.
+func (j *JSONWebToken) SerializeUnsecured(extraHeaders map[string]interface{}, detached bool) (string, error) {
+	return j.serialize(&unsecuredJWTSigner{extraHeaders}, detached)
+}
+
+func (j *JSONWebToken) serialize(signer jose.Signer, detached bool) (string, error) {
+	payloadBytes, err := j.marshalPayload()
+	if err != nil {
+		return "", fmt.Errorf("marshal JWT claims: %w", err)
+	}
+
+	jws := jose.NewJWS(j.Headers, nil, payloadBytes)
+	j.signature = jws.Signature()
+
+	return jws.SerializeCompact(signer, detached)
+}
+
+func (j *JSONWebToken) marshalPayload() ([]byte, error) {
+	return json.Marshal(j.Payload)
+}
+
+func parseJWS(jwt string, opts *parseOpts) (*JSONWebToken, error) {
+	jwsOpts := make([]jose.JWSParseOpt, 0)
+
+	if opts.detachedPayload != nil {
+		jwsOpts = append(jwsOpts, jose.WithJWSDetachedPayload(opts.detachedPayload))
+	}
+
+	jws, err := jose.ParseJWS(jwt, opts.sigVerifier, jwsOpts...)
+	if err != nil {
+		return nil, fmt.Errorf("parse JWT from compact JWS: %w", err)
+	}
+
+	return mapJWSToJWT(jws)
+}
+
+func mapJWSToJWT(jws *jose.JSONWebSignature) (*JSONWebToken, error) {
+	headers := jws.ProtectedHeaders
+
+	err := checkHeaders(headers)
+	if err != nil {
+		return nil, fmt.Errorf("check JWT headers: %w", err)
+	}
+
+	claims, err := toMap(jws.Payload)
+	if err != nil {
+		return nil, fmt.Errorf("read JWT claims from JWS payload: %w", err)
+	}
+
+	return &JSONWebToken{
+		Headers:   headers,
+		Payload:   claims,
+		signature: jws.Signature(),
+	}, nil
+}
+
+// New creates new JSON Web Token based on input claims.
+func New(claims interface{}) (*JSONWebToken, error) {
+	m, err := toMap(claims)
+	if err != nil {
+		return nil, fmt.Errorf("unmarshallable claims: %w", err)
+	}
+
+	return &JSONWebToken{
+		Payload: m,
+	}, nil
+}
+
+func checkHeaders(headers map[string]interface{}) error {
+	if _, ok := headers[jose.HeaderAlgorithm]; !ok {
+		return errors.New("alg header is not defined")
+	}
+
+	typ, ok := headers[jose.HeaderType]
+	if ok && typ != TypeJWT {
+		return errors.New("typ is not JWT")
+	}
+
+	cty, ok := headers[jose.HeaderContentType]
+	if ok && cty == TypeJWT {
+		return errors.New("nested JWT is not supported")
```

Comment:
what is this case?

Reply:
https://tools.ietf.org/html/rfc7519#section-5.2
I added a reference.